### PR TITLE
Invert dependency between Action_builder and Build system

### PR DESCRIPTION
**Pull request by jeremiedimino**
And remove the `Action_builder.Make_exec` function. After this PR, `Action_builder` becomes a monad implemented on top of `Build_system` and makes direct calls to `Build_system`. The next step will be to get rid of the variant and implement `Action_builder` in a more direct way.

As a result, the function `Build_system.build` is gone and replaced by `Action_builder.run`.

The removal of the functor means that we can no longer have an instance in `Print_rules` where we only collect dependencies without building them. This mode is important for `dune rules` so that we build the minimum required.

To keep `dune rules` doing the minimum amount of work after the removal of the functor, I added a global setting to put `Action_build` in "lazy mode". In this mode, building the dependencies recorded via the various `Action_builder.path`, ... functions is delayed. This allows to provide a function that only returns the dependencies without building them. This mode is only set by `dune rules`.

### CODE REVIEW

**Comment by aalekseyev** [comment]
I don't understand the need for this "lazy mode". It seems hard to reason about, in particular if you end up needing to run an action as a part of rule computation then the deps of that action will be broken, I think?

Can't we keep the functor, instead, or a parameter we can pass down?

**Comment by aalekseyev** [comment]
I'm not sure about this, but I find it suspicious that the `Action` constructor of action_builder now runs an action regardless of lazy mode, whereas previously I think those actions were not run in the corresponding mode.

To be fair, it's arguable either way which behavior is better, but I wanted to make sure I understand right and this is deliberate.

**Comment by jeremiedimino** [comment]
> I don't understand the need for this "lazy mode". It seems hard to reason about, in particular if you end up needing to run an action as a part of rule computation then the deps of that action will be broken, I think?

No, the `| Action` and `| Action_stdout` branch call `run` which ensure the dependencies are built independently of the mode.

> Can't we keep the functor, instead, or a parameter we can pass down?

That would work, but we might not get all the benefit we could get out of simplifying `Action_builder`, cf https://github.com/ocaml/dune/pull/4712

> I'm not sure about this, but I find it suspicious that the Action constructor of action_builder now runs an action regardless of lazy mode, whereas previously I think those actions were not run in the corresponding mode.

> To be fair, it's arguable either way which behavior is better, but I wanted to make sure I understand right and this is deliberate.

That's a fair point. I updated this PR to be conservative in this regard.

**Comment by jeremiedimino** [comment]
BTW, IIRC we only `Action_builder.action` for attaching anonymous actions to the `runtest` alias. This PR now prodive an explicit mechanism for that (see `Rules.Dir_rules.Alias_spec`). So we could in theory get rid of `Action_builder.action`, which would avoid the ambiguity. Although, we might still need it for the time being for the jenga backend.

**Comment by aalekseyev** [comment]
> That would work, but we might not get all the benefit we could get out of simplifying `Action_builder`, cf #4712

Hmm, I'm not a huge fan of this "mode" as it stands. Probably in part because nothing is documented about it, so I may be missing some things.

I also dislike the fact that we end up constructing "fake" values of `Dep.Facts.t` that don't actually carry any payload.

Let me try to reconstruct the line of reasoning here:

- it's nice to be able to use the same values of type `_ Action_builder.t` to work in both modes
- the two modes differ in how much facts we want to collect (and memoize), so we can't just use one implementation for both
- if we are trying to be "honest" and represent the action builder values with something like `'facts mode -> ('a * 'facts) Memo.Build.t`, then we can't implement `Action_builder.memoize` efficiently (at least we'd need to allocate two nodes instead of one for each memoize)
- we can simulate the "eager" mode on top of the "lazy" mode where necessary, so it's not a problem if we need both modes: we can just do everything in lazy mode and adapt where needed.

I can get on board with the change, but I think we should make things clearer and safer:

- We should clarify if it is safe to switch modes at will (then the "adapting" between modes has to be done by runtime inspection of returned values, similar to facts_or_deps), or does it have to be set at the beginning of the program and frozen (then we can make things more efficient by looking at the mode so we don't have to annotate every facts_or_deps with mode)
- Instead of `Dep.Facts.t` I think we could use `fact_or_dep Dep.Map.t` so that we don't carry around falsities anywhere.
- For efficient representation of `fact_or_dep` and expressive types we could look at the "runtime type equality" trick we have in jane (see runtime_type_equality library). Basically the trick is to annotate all values whose representation depends on mode with a type parameter, e.g. `fact_or_nothing`, and then have a global `val set_mode_eager : unit -> (fact_or_nothing, Fact.t) Type_equal.t` and `val set_mode_lazy : unit -> (fact_or_nothing, unit) Type_equal.t`.

FWIW, a simple thing to start with would be just an abstract type:
```
module Fact_or_dep : sig
  type t
end = struct 
  type t = Fact.t
end
```
whose accessors will be aware of the mode.

**Comment by aalekseyev** [comment]
By the way, if I understand it correctly, the `run` function in the eager mode actually discards the useful facts and recomputes them again, which seems not very efficient.

**Comment by jeremiedimino** [comment]
> By the way, if I understand it correctly, the run function in the eager mode actually discards the useful facts and recomputes them again, which seems not very efficient.

No, that's the opposite. In `Lazy` mode we discard the dummy `Fact.nothing` values and compute the real facts.

**Comment by aalekseyev** [comment]
> No, that's the opposite.

Got it, yeah, I somehow glossed over that pattern-match and assumed the code was run in all cases.

**Comment by jeremiedimino** [comment]
I pushed a commit that implements the idea of passing the mode as an argument as we discussed yesterday. It does seem cleaner this way

**Comment by aalekseyev** [comment]
I like the new approach. Do you have an understanding of the performance effects?
I'm expecting that creating two memo nodes instead of one is not going to help, but on the other hand the lookup in `Memo.Poly` is probably more expensive, and we're getting rid of that.

**Comment by jeremiedimino** [comment]
I expect the addition of a `mode` argument to `Action_builder.run` to have no effect and simply replace the functor argument. Then we are replacing the indirect calls to `Build_system` by direct calls in `Action_builder`. A few calls are now are behind a `match node with ...` but I would still expect it to be better than the indirect calls. So overall I would expect this PR to be either the same or better.

Good point about the creation of two memo nodes. We could use `_ Memo.Lazy.t Lazy.t` to avoid creating unnecessary nodes. In fact, we could even put a plain lazy inside `lazy_`?

**Comment by jeremiedimino** [comment]
I pushed the `Lazy.t`  wrapping at least in `Rule.memoize_thunk` as that seems strictly better.

**Comment by jeremiedimino** [comment]
> I expect the addition of a mode argument to Action_builder.run to have no effect and simply replace the functor argument.

The functor application had `[@@inlined]` so perhaps that's not true, but I would still expect adding an argument to `run` to have a negligible effect.

**Comment by aalekseyev** [comment]
@rgrinberg, this time the "invalid address format " happened in the CI (it was in a local build the last time)
Capturing the backtrace in case it has anything interesting:

```
  Error: exception Failure("invalid address format ")
  Raised at Stdlib.failwith in file "stdlib.ml", line 29, characters 17-33
  Called from Dune_rpc_private.Where.Make.get.of_file in file
    "otherlibs/dune-rpc/private/dune_rpc_private.ml", line 93, characters 15-35
  Called from Dune__exe__Rpc.establish_client_session.once in file
    "bin/rpc.ml", line 70, characters 16-42
  Called from Dune__exe__Rpc.retry_loop.loop in file "bin/rpc.ml", line 36,
    characters 15-22
```

I suspect this PR is not rebased, so doesn't have some of your fixes.

**Comment by aalekseyev** [comment]
It seems that in this PR we go out of our way to separate `Rule.thunk` from `Action_builder.t` even though conceptually the types are exactly the same (and as you said, we plan to make them exactly the same in the future when we "un-defunctionalize" the variant type).

This makes the downstream code somewhat arbitrarily in some places doing `Action_builder.run` and in others doing `.f`.

Since the types are (about to become) the same, the build system uses  this type, after all, so maybe we could be more upfront about it and aim for the following structure in the end?

`Action_builder_type`: abstract type definition, functions to inject and project (I know you don't like the `Action_builder0` convention, but that's maybe what I would call it)
`Rule` and `Build_system`: refer to `Action_builder_type`
`Action_builder`: contains a bunch of helper functions for constructing common values of `Action_builder_type.t`

I know at the moment it's awkward to make `Action_builder.t = Action_builder_type.t`, so some temporary adaptations may be needed.

File: bin/printenv.ml [hunk 1]
```diff
@@ -75,7 +75,11 @@ let term =
             User_error.raise
               [ Pp.text "Environment is not defined in install dirs" ])
       in
-      Build_system.run_exn (fun () -> Build_system.build request) >>| function
+      Build_system.run_exn (fun () ->
+          let open Memo.Build.O in
+          let+ res, _facts = Action_builder.run request Eager in
```
Comment (aalekseyev) [inline bin/printenv.ml:80]: Should we make this `Lazy`? (or leave a CR-someday)
It seems pretty unlikely that `Eager` is deliberate.

**Comment by aalekseyev** [comment]
@rgrinberg

aalekseyev> I suspect this PR doesn't have some of your fixes.

never mind, it does have the fixes I was thinking of.

**Review by aalekseyev — APPROVED** [review]
Looks good to me!
I'm not a fan of all those conversion between `Rule.thunk` and `Action_builder.t`, but  we can get rid of them in separate PRs.

**Comment by jeremiedimino** [comment]
I like the split of `Action_builder` you propose so that we can get rid of `Rule.thunk`, and I'm less fussy about `*0` modules nowadays. I think it makes sense here. Thanks for simplifying the action stuff!

**Pull request merged**